### ISSUE TITLE
Eliminate suid root across the images

### ISSUE THREAD

**Issue by RoyalOughtness** [secureblue/secureblue#296]

**Comment by fiftydinar** [comment]
Is it smart to alias `sudo` to `run0` to retain compatibility with some of the existing scripts?

**Comment by RoyalOughtness** [comment]
@fiftydinar which scripts? That's something to think about although it would cause issues for users who layer back `sudo`

**Comment by fiftydinar** [comment]
> @fiftydinar which scripts? That's something to think about although it would cause issues for users who layer back `sudo`

On top of my mind, some current rootful just scripts.

Bazzite relies on sudo for some of their hardware & software enablement scripts.

And possibly some others, from other sources.

**Comment by RoyalOughtness** [comment]
@fiftydinar good point. Either an alias or replacing `sudo` with `run0` via a script would be needed.

**Comment by secretmango** [comment]
I tried replacing sudo with doas once, and shutdown didnt work. So aliasing is for sure needed, even for system components.

**Comment by RoyalOughtness** [comment]
@firefoxlover doas isn't related to this issue. Also you can shutdown with systemd already via polkit with `systemctl shutdown`

**Comment by secretmango** [comment]
I dont know, I had an issue when shutting down in KDE plasma. True, systemd should work here. Will try it out when run0 is there, while an alias for sudo sounds like a good idea anyways.

**Comment by symbiogenesis** [comment]
run0 seems great. 256 is already in Fedora 41, and should be released soon. So having a PR in the works might be nice, so we can critique the implementation.

And documenting exactly why it is better than sudo or doas would be useful.

**Comment by RoyalOughtness** [comment]
> So having a PR in the works might be nice, so we can critique the implementation.

It's too soon. We won't be doing that until well after moving to f41 

> And documenting exactly why it is better than sudo or doas would be useful.

sudo and doas are both suid root binaries and are more trivially spoofable than a polkit prompt, used by run0. Also, run0's polkit integration makes daily driving a non-wheel user much less painful.

**Comment by KernelFreeze** [comment]
I tested removing `sudo` on the image `br-staging-41` and it works, you need to remove the file `/etc/dnf/protected.d/sudo.conf` that prevents removing it and then you can layer the removal with `rpm-ostree override remove sudo sudo-python-plugin`
![Image](https://github.com/user-attachments/assets/44dd1942-da01-444e-9090-640eac79a122)

**Comment by symbiogenesis** [comment]
Seems like the ultimate goal is that pkexec from polkit would be the only SUID-root program on the system. Replacing programs like sudo, passwd, etc when possible.

Maybe there could even be a way to enforce that? Like a whitelist-based approach to avoid pissing off devs and power users, and a friendly error message.

**Comment by RoyalOughtness** [comment]
> Seems like the ultimate goal is that pkexec from polkit would be the only SUID-root program on the system. 

No, `pkexec` is going away eventually too, there is no need for it with `run0`


> Maybe there could even be a way to enforce that? Like a whitelist-based approach to avoid pissing off devs and power users, and a friendly error message.

Please read the readme's hardening section before suggesting features. You're suggesting features we already have. We remove suid from all binaries except for those in the whitelist: https://github.com/secureblue/secureblue/blob/live/files/scripts/removesuid.sh

We then add back capabilities where needed, here: https://github.com/secureblue/secureblue/blob/live/files/system/usr/bin/setcapsforunsuidbinaries

**Comment by symbiogenesis** [comment]
That strips them from the image, but I meant that maybe there could be something like an SELinux policy or something where even at runtime it would disallow any programs from executing as SUID-root unless they are in a whitelist defined in /etc

This could be a shared whitelist that is also used by the removesuid.sh file.

**Comment by fiftydinar** [comment]
NOTE: Currently, `run0` always asks for the password for every root call in the script, unlike `sudo`.
So `--no-ask-password` doesn't work at the moment.

Has to do something with polkit:
https://github.com/systemd/systemd/issues/33366
https://github.com/polkit-org/polkit/issues/472

It might prompt the user multiple times for the password in some scripts, which is not ideal.

**Comment by secretmango** [comment]
https://github.com/systemd/systemd/labels/run

Here is the issuetracker for `run0` (they also use it for something called `systemd-run`, not sure what that is).

In an experimental stage it could be documented how to help test, but of course not replace it yet.

1. Alias `sudo=run0` and `pkexec=run0` at a system and `/etc/skel.d` level
2. Remove suid from `sudo` and `pkexec`
3. optional, remove sudo
  - Unprotect `sudo` with `run0 mv /etc/dnf/protected.d/sudo.conf /etc/dnf/protected.d/sudo.conf.disabled`
  - Remove `sudo` with `rpm-ostree override remove sudo`

Should this be a just command?

**Comment by fesdonomist** [comment]
Following this - I want to remove sudo from my own image at some point

**Comment by RoyalOughtness** [comment]
> https://github.com/systemd/systemd/labels/run
> 
> Here is the issuetracker for `run0` (they also use it for something called `systemd-run`, not sure what that is).

systemd-run is underneath run0, run0 is largely just a familiar alias to systemd-run

> In an experimental stage it could be documented how to help test, but of course not replace it yet.
> 
> 1. Alias `sudo=run0` and `pkexec=run0` at a system and `/etc/skel.d` level
> 2. Remove suid from `sudo` and `pkexec`
> 3. optional, remove sudo

Given that run0 isn't a drop-in replacement for sudo, I don't think this is a good option.

> * Unprotect `sudo` with `run0 mv /etc/dnf/protected.d/sudo.conf /etc/dnf/protected.d/sudo.conf.disabled`
> * Remove `sudo` with `rpm-ostree override remove sudo`
> 
> Should this be a just command?

Why would something that belongs in the image be a just command? Just commands are for things that *can't* be in the image for technical reasons

**Comment by RoyalOughtness** [comment]
based on my recent testing, the last binary that requires suid will be polkit's helper, and that will go away once this merges https://github.com/polkit-org/polkit/pull/501

**Comment by RoyalOughtness** [comment]
reopening, not fully complete
